### PR TITLE
Reviews screen - Show/hide `Mark all as read` bar button item based on unread notifications availability

### LINKED ISSUE #4295: Reviews: checkmark at top-right used to mark all reviews as read can be confusing
Steps to reproduce:

1. Switch to a store with no reviews.
1. Go to the Reviews screen.
1. Look for a checkmark at top-right and try to determine what it's for.
1. Switch to a store with at least one review.
1. Go to the Reviews screen.
1. Look for a checkmark at top-right and try to determine what it's for.

**Result:** the checkmark at top right is intended to mark all reviews as read, but it's a bit hard to figure that out if the store has no reviews or if the reviews have already been marked as read. 

No results view|Faded checkmark (all items already marked as read)
---|---
IMG_4167|IMG_4166

<sup>Tested with WCiOS 6.7.0.0 TestFlight beta on iPhone 6S iOS 14.6.</sup>

<sup>(internal reference: p5T066-2iC-p2#comment-8430 /ht @elibud)</sup>

Would it make sense to remove the checkmark from the no-results screen?

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #4295

### Description

This PR changes the behaviour of the  "Mark all as read" bar button item in `ReviewsViewController`.
- The button will be displayed only when there are unread notifications/reviews available. 
- Previously we were disabling the button if there are no unread notifications present.

### Changes
After making the necessary changes to ensure that the bar button item is shown/hidden as per the requirement, I observed that I had to do a small bit of refactoring and cleanup to make `ReviewsViewController` testable.
- Refactor `ReviewViewController` and `ReviewsViewModel` to improve testability.
- Write tests to validate the right bar button item's behaviour.
- Move localization strings to private enum. 
- Cleanup code.

### Testing instructions
1. Switch to a store with no reviews and navigate to the Review screen.
1. Observe that there is no right bar button item present.
1. Switch to a store with reviews and navigate to the Review screen.
1. Observe that the right bar button item is displayed when there are unread notifications present.
1. If there are no unread notifications present, please add a review to any of the products.
1. Refresh the Reviews screen and observe that the right bar button item is displayed.
1. Mark all unread notifications as read either by tapping on the right bar button item or by tapping and reading the review.
1. Refresh the Reviews screen and observe that the right bar button is not displayed.

Please note that the bar button item doesn't get refreshed properly due to another issue p91TBi-74J-p2. You have to switch to a different tab and navigate back to the Reviews screen to see the bar button item update. 

### Screenshots

| Review Screen state | Is Mark all as read button displayed? | Light | Dark|
| -------------  | ------------- | ------------- | ------------- |
| No unread notifications available | Mark all as read bar button item not displayed as there are no unread notifications. |   |   | 
| Unread notifications available | Mark all as read bar button item displayed. | |   | 
| No reviews | Mark all as read bar button item not displayed as there are no reviews available.  | |   | 

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/ViewRelated/Reviews/ReviewsViewControllerTests.swift [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// MARK: ReviewsViewController helpers
+//
+private extension ReviewsViewController {
+    func makeViewAppear() {
+        loadViewIfNeeded()
+        beginAppearanceTransition(true, animated: false)
+        endAppearanceTransition()
+    }
```
Comment (itsmeichigo) [inline WooCommerce/WooCommerceTests/ViewRelated/Reviews/ReviewsViewControllerTests.swift:51]: I'm curious: can you please explain the reason why these methods need to be called when testing the view controller?

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewModel.swift [hunk 2]
```diff
@@ -93,9 +130,9 @@ final class ReviewsViewModel {
 extension ReviewsViewModel {
     /// Prepares data necessary to render the reviews tab.
     ///
-    func synchronizeReviews(pageNumber: Int = Settings.firstPage,
-                            pageSize: Int = Settings.pageSize,
-                            onCompletion: (() -> Void)? = nil) {
+    func synchronizeReviews(pageNumber: Int,
+                            pageSize: Int,
+                            onCompletion: (() -> Void)?) {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewModel.swift:135]: Could you clarify the reason why we need to remove the default values here?

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewController.swift [hunk 3]
```diff
@@ -592,19 +585,36 @@ private extension ReviewsViewController {
...
+        }
+
+        enum EmptyState {
+            static let message = NSLocalizedString("Get your first reviews",
+                                                             comment: "Message shown in the Reviews tab if the list is empty")
+            static let detail = NSLocalizedString("Capture high-quality product reviews for your store.",
+                                                            comment: "Detailed message shown in the Reviews tab if the list is empty")
+            static let action = NSLocalizedString("Learn more",
+                                                            comment: "Title of button shown in the Reviews tab if the list is empty")
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewController.swift:618]: Thank you for re-organizing all the localization strings here! 👏

**Review by itsmeichigo — COMMENTED** [review]
Great job on the PR! 👍 

I have a few comments:
- It's so great that you reorganized a lot of stuff here to make everything cleaner and testable! However, I would suggest creating a separate issue for the refactor and a dedicated PR for it, to make it easier for reviewers to focus on the changes. Plus, such refactor can introduce regressions, so we'll need more testing. It's worth a separate point in `RELEASE_NOTES.md` file so that the team can do more smoke tests on the screen and find issues just in case.
- As for the main issue, I notice that we discussed another problem in #4295 that was missed out in this PR: Adam proposed [an update](https://github.com/woocommerce/woocommerce-ios/issues/4295#issuecomment-888041823) to the bar button item, which I think is really nice since we all agree that the check button isn't very obvious in terms of meaning. Do you think you can add it to this PR?
- Regarding the issue updating the bar button item, I commented in the internal P2 post that I could not reproduce the issue on `trunk`. Can you double-check to make sure that it was not introduced in this PR?
- Thank you for adding the unit tests for the `ReviewsViewController`! I've seen folks writing tests for view controllers using `Reflection` (for example [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/WooCommerceTests/ViewRelated/ReusableViews/EmptyStateViewController/EmptyStateViewControllerTests.swift)). Do you think this can be applied to the new tests too?

File: WooCommerce/WooCommerceTests/ViewRelated/Reviews/ReviewsViewControllerTests.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/WooCommerceTests/ViewRelated/Reviews/ReviewsViewControllerTests.swift:51]: These methods simulate the act of presenting a view controller. (The view controller's viewDidLoad and viewWillAppear methods are called.)

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewModel.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewModel.swift:135]: 1. I introduced a protocol that defines this method and we cannot have methods with default parameters in protocols.
1. There was no need for the default parameter at all, except in the `ReviewViewModelTests`. 

So, I decided to remove those and pass in default values from tests, just like what other test classes follow. Kindly let me know if you think that this isn't alright.

**Comment by selanthiraiyan** [comment]
Closing this PR in order to,

1. Break this down into smaller PRs to make code review easier. Currently, this PR is mixed with refactoring, feature and cleanup related code which makes the code review process difficult.
1. Add the missed feature which is replacing the checkmark icon with the menu button and adding an action sheet.

Thank you @itsmeichigo for your inputs! 🙇

**Pull request closed without merging**